storage: fixes TestReplicateQueueDownReplicate

BramGruneir opened the pull request:
Fixes #12943.
Fixes #11645.
Fixes #11640.

BramGruneir committed:
storage: fixes TestReplicateQueueDownReplicate

A few changes were required to fix this test:
- capture 2 potential errors during up-replication that don't affect the outcome or desired behavior of the test
- ensure that each up-replication occurs individually and will be retried when possible
- looking at the range descriptor instead of just counting the replicas directly on the stores so replicas waiting to be GCed are not mistakenly counted

Fixes #12943.
Fixes #11645.
Fixes #11640.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15340)
<!-- Reviewable:end -->

tamird commented:
How about a description of the problem in the commit message?

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 108 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-Kiaj8OHwtoL5hEbpYMT:-Kiaj8OHwtoL5hEbpYMU:b-yilddd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf91df2421a4ee57ff5f67fc82ce1ba7aadefd75/pkg/storage/replicate_queue_test.go#L108)):*
> ```Go
> // notice over-replicated ranges and remove replicas from them.
> //
> // The goal of this test is to ensure that down replication occurs correctly
> ```

I don't think we need this comment in two places. On the implementation seems more appropriate.

---

*[pkg/storage/replicate_queue_test.go, line 140 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-Kiailfyijl1FzsgL2wk:-Kiailfyijl1FzsgL2wl:b-xrv63q) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf91df2421a4ee57ff5f67fc82ce1ba7aadefd75/pkg/storage/replicate_queue_test.go#L140)):*
> ```Go
> 		// retry is in order.
> 		testutils.SucceedsSoon(t, func() error {
> 			if _, err := tc.AddReplicas(testKey, roachpb.ReplicationTarget{
> ```

Looks like `AddReplicas` takes multiple replicas. Does that not work?

---

*[pkg/storage/replicate_queue_test.go, line 149 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-KiajIKHpw4SFLPJMBbY:-KiajIKHpw4SFLPJMBbZ:b-ehb2ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf91df2421a4ee57ff5f67fc82ce1ba7aadefd75/pkg/storage/replicate_queue_test.go#L149)):*
> ```Go
> 				// There's a chance that the replica has already been removed
> 				// but not GCed yet. This is the expected error for that case.
> 				!testutils.IsError(err, "snapshot intersects with existing range") {
> ```

this looks like the only behaviour change here; is that right? I must be missing something because the old code also called `tc.AddReplicas`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-:-KiaiLY07twngkeLoIoY:b-ctlttn)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Added more details in the commit message.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/replicate_queue_test.go, line 108 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-Kiaj8OHwtoL5hEbpYMT:-Kiaprvfkb8a923xpQlx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf91df2421a4ee57ff5f67fc82ce1ba7aadefd75/pkg/storage/replicate_queue_test.go#L108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I don't think we need this comment in two places. On the implementation seems more appropriate.
</blockquote></details>

Done.

---

*[pkg/storage/replicate_queue_test.go, line 140 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-Kiailfyijl1FzsgL2wk:-KialHY7BZvIk4g3Ph3P:b-25lf0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf91df2421a4ee57ff5f67fc82ce1ba7aadefd75/pkg/storage/replicate_queue_test.go#L140)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Looks like `AddReplicas` takes multiple replicas. Does that not work?
</blockquote></details>

Sadly, no it doesn't work.  It will fail with one of the errors every time and never up-replicate. 

---

*[pkg/storage/replicate_queue_test.go, line 149 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-KiajIKHpw4SFLPJMBbY:-Kiansk0Umtg7_FBC3SM:b8p1yjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf91df2421a4ee57ff5f67fc82ce1ba7aadefd75/pkg/storage/replicate_queue_test.go#L149)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this looks like the only behaviour change here; is that right? I must be missing something because the old code also called `tc.AddReplicas`.
</blockquote></details>

Correct, the main behaviour change is this.  Also, the old replicaCount function counted all replicas on a store, but didn't look into if the replica was scheduled to be GCed or not, relying also on the replica GC queue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-:-KiasCUjjbBjmQdrr-2i:b-laghti)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-:-Kib8VlcYDSZ6mrm-O95:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/replicate_queue_test.go, line 144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-KibAZBzaCHmgH1o9Jpo:-KibAZBzaCHmgH1o9Jpp:b-lv14a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/02796c6da553ed7a441c43c237d115816a30ce43/pkg/storage/replicate_queue_test.go#L144)):*
> ```Go
> 				// This error occurs when the node already has a replica for
> 				// this range.
> 				!testutils.IsError(err, "which is already present") &&
> ```

could you keep the longer matcher that was previously here?

also, I don't think there's any need for this nil check.

```
_, err := tc.AddReplicas(testKey, roachpb.ReplicationTarget{
	NodeID:  nodeID,
	StoreID: server.GetFirstStoreID(),
})
if testutils.IsError(err, strings.Join([]string{"unable to add replica .* which is already present", storage.IntersectingSnapshotMsg}, "|")) {
  return nil
}
return err
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-KibAZBzaCHmgH1o9Jpo:-Kicavk5eaL8fUym7V37:b-vbwfyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/02796c6da553ed7a441c43c237d115816a30ce43/pkg/storage/replicate_queue_test.go#L144)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

could you keep the longer matcher that was previously here?

also, I don't think there's any need for this nil check.

```
_, err := tc.AddReplicas(testKey, roachpb.ReplicationTarget{
	NodeID:  nodeID,
	StoreID: server.GetFirstStoreID(),
})
if testutils.IsError(err, strings.Join([]string{"unable to add replica .* which is already present", storage.IntersectingSnapshotMsg}, "|")) {
  return nil
}
return err
```
</blockquote></details>

I know this is faster due to using only a single regex, but I find it hard to parse and more difficult to comment on.  I've cleaned it up, removed the nil, and put back in the longer matcher.  But I'm leaving the two checks instead of the one for readability.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/replicate_queue_test.go, line 144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-KibAZBzaCHmgH1o9Jpo:-KiecmT26JQonj3Blv-b:b-dqyznx) ([raw file](https://github.com/cockroachdb/cockroach/blob/02796c6da553ed7a441c43c237d115816a30ce43/pkg/storage/replicate_queue_test.go#L144)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I know this is faster due to using only a single regex, but I find it hard to parse and more difficult to comment on.  I've cleaned it up, removed the nil, and put back in the longer matcher.  But I'm leaving the two checks instead of the one for readability.
</blockquote></details>

If you're looking for comment room, consider:
```
allowedErrs := strings.Join([]string{
  // first comment.
  "unable to add replica .* which is already present",
  // second comment.
  storage.IntersectingSnapshotMsg,
}, "|")
if testutils.IsError(err, allowedErrs) {
  return nil
}
return err
```

if you make this change, also suggest you remove the words "first" and "second" since they're asking for rot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-:-KiedEjPMO4aTdaNhZsl:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replicate_queue_test.go, line 144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340#-KibAZBzaCHmgH1o9Jpo:-KiehlJyQ9KXjM8ndsuI:bfrutol) ([raw file](https://github.com/cockroachdb/cockroach/blob/02796c6da553ed7a441c43c237d115816a30ce43/pkg/storage/replicate_queue_test.go#L144)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

If you're looking for comment room, consider:
```
allowedErrs := strings.Join([]string{
  // first comment.
  "unable to add replica .* which is already present",
  // second comment.
  storage.IntersectingSnapshotMsg,
}, "|")
if testutils.IsError(err, allowedErrs) {
  return nil
}
return err
```

if you make this change, also suggest you remove the words "first" and "second" since they're asking for rot.
</blockquote></details>

I did one better and moved the join outside of the loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15340)*
<!-- Sent from Reviewable.io -->
